### PR TITLE
Add CertificatesExpiration to cluster status 

### PR DESCRIPTION
**Pull request by moelsayed**
rancher/rancher#19103

### CODE REVIEW

File: pkg/controllers/management/certsexpiration/certsexpiration.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+		certsExpInfo[certName] = info
+	}
+	cluster.Status.CertificatesExpiration = certsExpInfo
+	return c.ClusterClient.Update(cluster)
+
+}
+
+func (c Controller) getClusterCertificateBundle(clusterName string) (map[string]pki.CertificatePKI, error) {
+	// cluster has a state file ?
+	store := clusterprovisioner.NewPersistentStore(c.NamespaceClient, c.CoreV1)
```
Comment (kinarashah) [inline pkg/controllers/management/certsexpiration/certsexpiration.go]: why are we getting a new store every time? would it help to just have it in controller definition?

File: pkg/controllers/management/certsexpiration/certsexpiration.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+		return nil, err
+	}
+	for certName, certObj := range certBundle {
+		info, err := getCertExpiration(certObj.CertificatePEM)
+		if err != nil {
+			continue
+		}
+		certsExpInfo[certName] = info
+	}
+	cluster.Status.CertificatesExpiration = certsExpInfo
```
Comment (kinarashah) [inline pkg/controllers/management/certsexpiration/certsexpiration.go]: we should update this only if certsExpInfo has changed?

File: pkg/controllers/management/certsexpiration/certsexpiration.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	return certs, nil
+}
+
+func getRKECurrentStateFromStore(store cluster.PersistentStore, clusterName string) (*rkecluster.State, error) {
+	cluster, err := store.Get(clusterName)
+	if err != nil {
+		return nil, err
+	}
+	var fullState rkecluster.FullState
+	stateStr, ok := cluster.Metadata["fullState"]
```
Comment (kinarashah) [inline pkg/controllers/management/certsexpiration/certsexpiration.go]: just a question, when will `fullState` not be present? in that case, do we need to look for it elsewhere?

File: pkg/controllers/management/certsexpiration/certsexpiration.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return &fullState.CurrentState, nil
+}
+
+func getCertsFromUserCluster(c *config.UserContext) (map[string]pki.CertificatePKI, error) {
+	certs := map[string]pki.CertificatePKI{}
+
+	secrets, err := c.Core.Secrets("kube-system").List(metav1.ListOptions{})
```
Comment (kinarashah) [inline pkg/controllers/management/certsexpiration/certsexpiration.go]: can we use lister here?

File: pkg/controllers/management/certsexpiration/certsexpiration.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		return nil, err
+	}
+	for _, s := range secrets.Items {
+
+		if strings.HasPrefix(s.GetName(), "kube-") &&
+			s.Type == corev1.SecretTypeOpaque {
+			name := s.GetName()
+			secret, err := c.Core.Secrets("kube-system").Get(name, metav1.GetOptions{})
```
Comment (kinarashah) [inline pkg/controllers/management/certsexpiration/certsexpiration.go]: same here, Controller.Lister.Get

File: pkg/controllers/management/certsexpiration/certsexpiration.go (hunk 2 above)
Comment (kinarashah) [reply pkg/controllers/management/certsexpiration/certsexpiration.go]: also deepcopy before update

File: pkg/controllers/management/certsexpiration/certsexpiration.go (hunk 3 above)
Comment (moelsayed) [reply pkg/controllers/management/certsexpiration/certsexpiration.go]: No, this is only available in 2.2.x clusters. If it's not found, it's 2.0.x or 2.1.x cluster, in that case we get the certificate bundle from the user cluster directly.

File: pkg/controllers/management/certsexpiration/certsexpiration.go (hunk 4 above)
Comment (moelsayed) [reply pkg/controllers/management/certsexpiration/certsexpiration.go]: Unfortunately, no. It's a UserContext, in this case we hit the user cluster directly.

**Pull request closed without merging**